remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,562 @@
+          in_offset_map = true;
+        } else {
+          if (ptr->second != offset) {
+            // bail if the offsets don't match
+            return VisitResult::Error;
+          }
+        }
+      }
+    }
+  }
+  if (in_offset_map && in_exclude_set) {
+    // fail if some operands are excluded and others are state offsets
+    return VisitResult::Error;
+  } else if (in_offset_map) {
+    LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+    offset_map.insert({&I, offset});
+  } else if (in_exclude_set) {
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    exclude.insert(&I);
+  }
+  return VisitResult::Progress;
+}
+
+// Return the scope of the given instruction.
+llvm::MDNode *GetScopeFromInst(llvm::Instruction &I) {
+  llvm::AAMDNodes N;
+  I.getAAMetadata(N);
+  return N.Scope;
+}
+
+// Remove all dead stores.
+void RemoveDeadStores(const std::unordered_set<llvm::Instruction *> &dead_stores) {
+  for (auto *store : dead_stores) {
+    store->eraseFromParent();
+  }
+}
+
+LiveSetBlockVisitor::LiveSetBlockVisitor(const ScopeMap &scope_to_offset_, const llvm::FunctionType *lft_)
+  : scope_to_offset(scope_to_offset_),
+    curr_wl(),
+    next_wl(),
+    block_map(),
+    to_remove(),
+    lft(lft_) {}
+
+// Get every terminating basic block from the function `func`.
+void LiveSetBlockVisitor::AddFunction(llvm::Function &func) {
+  for (auto &block : func) {
+    // The machines rose from the ashes of the nuclear fire....
+    if (block.getTerminator()) {
+      curr_wl.push_back(&block);
+      // add block to map with initial entry set
+      block_map.insert({&block, LiveSet()});
+    }
+  }
+}
+
+// Visit the basic blocks in the worklist and update the block_map.
+void LiveSetBlockVisitor::Visit(void) {
+  bool progress = true;
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty() && progress) {
+    LOG(INFO) << "LSBV: " << curr_wl.size() << " blocks in worklist";
+    progress = false;
+    for (auto block : curr_wl) {
+      if (VisitBlock(block)) {
+        progress = true;
+        // Updates have been made; add predecessors to next rotation
+        for (llvm::BasicBlock *pred : predecessors(block)) {
+          next_wl.push_back(pred);
+        }
+      }
+      //LOG(INFO) << "LSBV: " << live->count() << " live slots";
+    }
+    // update curr_wl to the next set of insts
+    curr_wl.swap(next_wl);
+    // reset upcoming insts
+    next_wl.clear();
+  }
+  LOG(INFO) << "Found " << to_remove.size() << " dead stores to remove";
+  RemoveDeadStores(to_remove);
+}
+
+bool LiveSetBlockVisitor::VisitBlock(llvm::BasicBlock *B) {
+  //block_map[B] = *init;
+  LiveSet live;
+  for (auto inst_it = B->rbegin(); inst_it != B->rend(); ++inst_it) {
+    auto inst = &*inst_it;
+    if (llvm::isa<llvm::ReturnInst>(inst) || llvm::isa<llvm::UnreachableInst>(inst)) {
+      live.set();
+    } else if (llvm::isa<llvm::BranchInst>(inst)) {
+      // update live from successors
```

Comment:
Oh we also want to handle `llvm::SwitchInst` and other types of terminators: https://llvm.org/docs/LangRef.html#terminator-instructions